### DISCUSSION TITLE
Can we make Fastapi endpooints with path variable optional ?

### DISCUSSION THREAD

**Discussion by aru88** [fastapi/fastapi#9516] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.get(
    '/api/v1/details/userName={user_name}',
    name="data:get-user-details",
    tags=["USerManagement"],
    status_code=status.HTTP_200_OK,
    summary="Get userdetails from database table"
)
async def get_user_details(
    userName: Optional[str] = None,
    data_service: Depends = Depends(get_data_service)
):
```

### Description

Hello,

i have to create endpoint with the path variable and query parameters.
both path variable and query parameters are optional.
so endpoint definitions are
/api/v1/details
/api/v1/details/name=user_name
api/v1/details?group=private

is it possible to design single  end point to achieve this ?

Thanks& regards

### Operating System

Linux

### FastAPI Version

0.75.1

### Python Version

3.8

**Comment by chrisK824** [comment]
Hi @aru88 ,  based on your endpoints:

/api/v1/details
/api/v1/details/name=user_name
api/v1/details?group=private

I think you just need two Query parameters.
Something like this should do it:

```
@router.get(
    '/api/v1/details',
    name="data:get-user-details",
    tags=["USerManagement"],
    status_code=status.HTTP_200_OK,
    summary="Get userdetails from database table"
)
async def get_user_details(
    userName:  str = Query(default=None),
    group: str = Query(default=None)
    data_service: Depends = Depends(get_data_service)
):
    print(userName)
    print(group)
```

Hope that helps!

**Reply by aru88** [reply]
Thanks for your response. is it possible i can use userName as path parameter like /api/v1/details/name={user_name}

**Reply by chrisK824** [reply]
Not sure what you want to achieve @aru88 , the `=` sign there suggests you need a Query parameter and not a Path one.
If you want a path parameter this could be done like that:

Endpoint:   /api/v1/details/{user_name}

```
@router.get(
    '/api/v1/details/{user_name}',
    name="data:get-user-details",
    tags=["USerManagement"],
    status_code=status.HTTP_200_OK,
    summary="Get userdetails from database table"
)
async def get_user_details(
    userName:  str ,
    data_service: Depends = Depends(get_data_service)
):
    print(userName)
    print(group)
```

**Reply by aru88** [reply]
we want to implement the Matrix parameter somewhat like  https://www.w3.org/DesignIssues/MatrixURIs.html
something like /nodes/userLabel={userLabel}/connected-nodes

**Comment by GITHub8803** [comment]
Not me Robby williams hacking my shit

**Comment by djimontyp** [comment, marked as answer]
@aru88 You cannot use default options in Path. A Path is needed.

fastapi.params.py > 14 from source
```
class Path(Param):
    in_ = ParamTypes.path

    def __init__(
        self,
        default: Any = ...,
        *,
        alias: Optional[str] = None,
        title: Optional[str] = None,
        description: Optional[str] = None,
        gt: Optional[float] = None,
        ge: Optional[float] = None,
        lt: Optional[float] = None,
        le: Optional[float] = None,
        min_length: Optional[int] = None,
        max_length: Optional[int] = None,
        regex: Optional[str] = None,
        example: Any = Undefined,
        examples: Optional[Dict[str, Any]] = None,
        deprecated: Optional[bool] = None,
        include_in_schema: bool = True,
        **extra: Any,
    ):
        assert default is ..., "Path parameters cannot have a default value"
```
 
I recommend a separate endpoint.
 
```
from typing import Optional

import uvicorn as uvicorn
from fastapi import FastAPI, HTTPException
from starlette import status

app = FastAPI()

db = {
    "alice": "alice",
    "bob": "bob",
}


@app.get("/api/v1/details")
async def get_user_details(group: Optional[str] = None):
    return db


@app.get("/api/v1/details/{userName}")
async def get_user_details(userName: str, group: Optional[str] = None):
    if user := db.get(userName):
        return user

    raise HTTPException(status_code=status.HTTP_404_NOT_FOUND)


if __name__ == "__main__":
    uvicorn.run(app)

```

**Comment by jgould22** [comment]
You may be able to achieve what you want with a custom path converter from within Starlette.

https://www.starlette.io/routing/

I am unsure though if that stuff can be exposed in fastapi or used, you will have to do more research or try it and see.
